### PR TITLE
frontend: fix for authentication error message flickering issue

### PR DESCRIPTION
**Pull request by acodegirl**
Jira # PAY2-66

### CODE REVIEW

File: assets/src/auth/pages/RestrictedPage.tsx [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 
 interface Props {
   children: JSX.Element;
 }
 
 const RestrictedPage: React.FC<Props> = ({ children }) => {
-  const user = usePayUser();
+  const { isInitializing, isAuthenticated } = useAuth0();
+
+  if (isInitializing) return <Loader message="Loading " />;
```
Comment (jonathaningram) [inline assets/src/auth/pages/RestrictedPage.tsx]: @acodegirl is your prettier running? this `if` should be surrounded with curly braces?

File: assets/src/platform-admin/pages/RestrictedPage.tsx [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
 
 interface Props {
   children: JSX.Element;
 }
 
 const RestrictedPage: React.FC<Props> = ({ children }) => {
-  const user = usePayUser();
+  const { isInitializing, isAuthenticated } = useAuth0();
+
+  if (isInitializing) return <Loader message="Loading " />;
```
Comment (jonathaningram) [inline assets/src/platform-admin/pages/RestrictedPage.tsx]: @acodegirl as above

File: assets/src/platform-admin/pages/RestrictedPage.tsx [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
 
-  if (user && user.platformAdmin) {
```
Comment (jonathaningram) [inline assets/src/platform-admin/pages/RestrictedPage.tsx]: @acodegirl this restricted page component is inside the `platform-admin` app. by changing this condition, are you saying that any user is allowed to access platform admin pages if they are authenticated? or is the check for `user.platformAdmin === true` occurring elsewhere now?

File: assets/src/auth/pages/RestrictedPage.tsx (hunk 1 above)
Comment (jonathaningram) [reply assets/src/auth/pages/RestrictedPage.tsx]: @acodegirl you have a trailing space in the `message`

File: assets/src/platform-admin/pages/RestrictedPage.tsx (hunk 2 above)
Comment (jonathaningram) [reply assets/src/platform-admin/pages/RestrictedPage.tsx]: @acodegirl you have a trailing space in the `message`

File: assets/src/layout/Header/index.tsx [hunk 4]
```diff
@@ -191,7 +191,7 @@ const Header: React.FC = () => {
           </PrimaryContainer>
         </PrimaryHeader>
 
-        {isAuthenticated ? (
+        {isAuthenticated && user ? (
```
Comment (jonathaningram) [inline assets/src/layout/Header/index.tsx]: @acodegirl does this logic make sense? "if authenticated and if user". If authenticated I would expect there to always be a user, so logically I'm not sure the right hand side of this operation should be required?

File: assets/src/users/UserContext.tsx [hunk 5]
```diff
@@ -48,16 +55,21 @@ const useAuthUser = () => {
   const { user } = useAuth0();
   return user;
 };
-const usePayUser = () => React.useContext(PayUserContext);
+const usePayUser = () => React.useContext(PayUserContext)!;
```
Comment (jonathaningram) [inline assets/src/users/UserContext.tsx]: @acodegirl this `!` makes me nervous. what led to this being required? I wonder if there's a way to not use this?

File: assets/src/users/UserContext.tsx [hunk 6]
```diff
@@ -48,16 +55,21 @@ const useAuthUser = () => {
...
+  let isLoading = false;
+  const { isInitializing } = useAuth0();
+  const { loading, data } = useCheckAuthQuery({});
+  if (isInitializing || loading) {
+    isLoading = true;
+  }
+  const user = data && data.me ? fromGQLUser(data.me) : undefined;
   return (
-    <PayUserContext.Provider value={value}>{children}</PayUserContext.Provider>
+    <PayUserContext.Provider value={{ user, isLoading }}>
```
Comment (jonathaningram) [inline assets/src/users/UserContext.tsx]: @acodegirl perhaps it is better to set the value to `{ user, isLoading: isInitializing || loading }` directly. Then you can remove the mutable `isLoading` local variable state.

File: assets/src/layout/Header/index.tsx (hunk 4 above)
Comment (acodegirl) [reply assets/src/layout/Header/index.tsx]: @jonathaningram - Soon after authentication, its taking a bit of time to fetch pay user data. But meanwhile couple of graphql requests are sent to fetch service related info which were failing and sent again after receiving successful user information. This is the reason i added this check to ensure graphql requests are sent only after fetching user info. Please refer  the above screenshot attached by Peter.

Comment (jonathaningram) [reply assets/src/layout/Header/index.tsx]: @acodegirl ok no problem thanks

File: assets/src/platform-admin/pages/RestrictedPage.tsx [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
 interface Props {
   children: JSX.Element;
 }
 
 const RestrictedPage: React.FC<Props> = ({ children }) => {
-  const user = usePayUser();
+  const { isLoading, user } = usePayUser();
+
+  if (isLoading) {
+    return <Loader message="Loading" />;
```
Comment (jonathaningram) [inline assets/src/platform-admin/pages/RestrictedPage.tsx]: @acodegirl minor, but the `message` prop is already `Loading` so no need to use the prop here.

File: assets/src/users/UserContext.tsx [hunk 8]
```diff
@@ -35,11 +40,11 @@ const AuthUserProvider: React.FC<{}> = ({ children }) => {
   );
 };
 
-const PayUserContext = React.createContext<User | undefined>(undefined);
+const PayUserContext = React.createContext<Partial<PayUserContext>>({});
 const UserContext = React.createContext<
   | {
       authUser: Auth0User;
-      payUser: User;
+      payUser: Partial<PayUserContext>;
```
Comment (jonathaningram) [inline assets/src/users/UserContext.tsx]: @acodegirl is there a specific behaviour you're going for, or is this mostly about making TS be quiet? I checked out the branch locally and is this a better solution for these two contexts?

```ts
const PayUserContext = React.createContext<PayUserContext>({
  isLoading: false
});
const UserContext = React.createContext<
  | {
      authUser: Auth0User;
    }
  | undefined
>(undefined);
```

Note: I removed `payUser` from the `UserContext`. It's not used anywhere, so I don't think it needs to be in there, at least not for now?

**Pull request merged**